File: src/game/graphics/texture.rs
/*
 * Copyright (c) 2021 LambdAurora <aurora42lambda@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::ffi::c_void;
use std::path::Path;

use stb_image::image::{Image, LoadResult};

/**
 * Represents a texture.
 */
#[derive(Clone)]
pub struct Texture {
    id: u32,
    width: usize,
    height: usize,
}

impl Texture {
    pub fn new(image: Image<u8>) -> Result<Texture, String> {
        let mut id: u32 = 0;
        unsafe { gl::GenTextures(1, &mut id); }
        if id == 0 {
            return Err(String::from("Could not generate the OpenGL texture."));
        }

        let mut texture = Texture { id, width: image.width, height: image.height };
        texture.upload(image);
        Ok(texture)
    }

    /// Loads the texture from the given path.
    pub fn load(path: &Path, flip: bool) -> Result<Texture, String> {
        unsafe { stb_image::stb_image::bindgen::stbi_set_flip_vertically_on_load(flip as i32); }
        Texture::new(load_image(path)?)
    }

    pub fn get_format(image: &Image<u8>) -> u32 {
        match image.depth {
            1 => gl::RED,
            2 => gl::RG,
            3 => gl::RGB,
            4 => gl::RGBA,
            _ => 0
        }
    }

    /// Binds the texture.
    pub fn bind(&self) {
        unsafe { gl::BindTexture(gl::TEXTURE_2D, self.id); }
    }

    /// Uploads the given image to the texture.
    pub fn upload(&mut self, image: Image<u8>) {
        self.bind();

        self.width = image.width;
        self.height = image.height;

        let format = Texture::get_format(&image);
        unsafe {
            gl::TexImage2D(gl::TEXTURE_2D, 0, format as i32,
                           self.width as i32, self.height as i32,
                           0, format, gl::UNSIGNED_BYTE, image.data.as_ptr() as *const c_void);
            gl::GenerateMipmap(gl::TEXTURE_2D);

            gl::TexParameteri(gl::TEXTURE_2D, gl::TEXTURE_WRAP_S, gl::REPEAT as i32);
            gl::TexParameteri(gl::TEXTURE_2D, gl::TEXTURE_WRAP_T, gl::REPEAT as i32);
            gl::TexParameteri(gl::TEXTURE_2D, gl::TEXTURE_MIN_FILTER, gl::NEAREST_MIPMAP_NEAREST as i32);
            gl::TexParameteri(gl::TEXTURE_2D, gl::TEXTURE_MAG_FILTER, gl::NEAREST as i32);
        }
    }
}

/// Unbinds texture.
pub fn unbind() {
    unsafe { gl::BindTexture(gl::TEXTURE_2D, 0); }
}

pub fn load_image(path: &Path) -> Result<Image<u8>, String> {
    match stb_image::image::load(path) {
        LoadResult::ImageU8(img) => Ok(img),
        LoadResult::ImageF32(_) => Err(String::from("Does not support ImageF32")),
        LoadResult::Error(err) => Err(err)
    }
}

File: src/main/java/dev/lambdaurora/res_errare/render/texture/Texture.java
/*
 * Copyright (c) 2021 LambdAurora <aurora42lambda@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package dev.lambdaurora.res_errare.render.texture;

import dev.lambdaurora.res_errare.system.GL;
import dev.lambdaurora.res_errare.system.OpenGLIdProvider;
import jdk.incubator.foreign.MemoryAddress;

/**
 * Represents a texture.
 *
 * @param <T> the accepted texture targets
 */
public interface Texture<T extends OpenGLIdProvider> extends AutoCloseable {
	TextureType type();

	/**
	 * {@return the OpenGL identifier of this texture}
	 */
	int id();

	/**
	 * Binds the texture for use.
	 */
	default void bind() {
		GL.get().bindTexture(this.type(), this.id());
	}

	/**
	 * Unbinds the texture.
	 */
	default void unbind() {
		unbind(this.type());
	}

	/**
	 * Uploads the given image to the texture.
	 *
	 * @param target the texture target
	 * @param level the level
	 * @param internalFormat the internal format used by OpenGL for this texture
	 * @param width the width of the texture
	 * @param height the height of the texture
	 */
	default void initEmpty(T target, int level, InternalFormat internalFormat, int width, int height) {
		this.bind();
		GL.get().texImage2D(target, level, internalFormat, width, height, Image.Format.ARGB, internalFormat.typeId(), MemoryAddress.NULL);
	}

	/**
	 * Uploads the given image to the texture.
	 *
	 * @param target the texture target
	 * @param level the level
	 * @param image the image to upload
	 */
	default void upload(T target, int level, Image image) {
		this.upload(target, level, image.format().internalFormat(), image);
	}

	/**
	 * Uploads the given image to the texture.
	 *
	 * @param target the texture target
	 * @param level the level
	 * @param internalFormat the internal format used by OpenGL for this texture
	 * @param image the image to upload
	 */
	default void upload(T target, int level, InternalFormat internalFormat, Image image) {
		GL.get().texImage2D(target, level, internalFormat, image);
	}

	default BoundImageTexture<T> bindImageTexture(int unit, int level, GL.Access access, InternalFormat format) {
		GL.get().bindImageTexture(unit, this.id(), level, false, 0, access, format);
		return new BoundImageTexture<>(unit, this, access, format);
	}

	default BoundImageTexture<T> bindImageTexture(int unit, int level, int layer, GL.Access access, InternalFormat format) {
		GL.get().bindImageTexture(unit, this.id(), level, true, layer, access, format);
		return new BoundImageTexture<>(unit, this, access, format);
	}

	default <V> void setParameter(TextureParameter<V> parameter, V value) {
		switch (parameter.type()) {
			case INTEGER -> GL.get().texParameteri(this.type(), parameter, parameter.getIntValue(value));
			case FLOAT -> GL.get().texParameterf(this.type(), parameter, parameter.getFloatValue(value));
		}
	}

	default void generateMipmap() {
		GL.get().generateMipmap(this.type());
	}

	/**
	 * Deletes the texture. Once deleted the texture should not be used anymore.
	 */
	@Override
	default void close() {
		GL.get().deleteTextures(this.id());
	}

	static void unbind(TextureType type) {
		GL.get().bindTexture(type, 0);
	}

	enum InternalFormat implements OpenGLIdProvider {
		R8UI(0x8232, GL.GL11.UNSIGNED_BYTE),
		R32F(0x822e, GL.GL11.FLOAT),
		RGB8UI(0x8d7d, GL.GL11.UNSIGNED_BYTE),
		RGB32F(0x8815, GL.GL11.FLOAT),
		RGBA8UI(0x8d7c, GL.GL11.UNSIGNED_BYTE),
		RGBA32F(0x8814, GL.GL11.FLOAT);

		private final int glId;
		private final int typeId;

		InternalFormat(int glId, int typeId) {
			this.glId = glId;
			this.typeId = typeId;
		}

		@Override
		public int glId() {
			return this.glId;
		}

		public int typeId() {
			return this.typeId;
		}
	}
}
